src/dp/distributions.rs

```diff
@@ -0,0 +1,323 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+//
+// This file incorporates work covered by the following copyright and
+// permission notice:
+//
+//   Copyright 2020 Thomas Steinke
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+//   The following code is adapted from the opendp implementation to reduce dependencies:
+//       https://github.com/opendp/opendp/blob/main/rust/src/traits/samplers/cks20
+
+use num_bigint::{BigInt, BigUint};
+use num_traits::{One, Zero};
+use rand::{distributions::Distribution, distributions::Uniform, Rng};
+
+use super::{
+    BigURational, DifferentialPrivacyBudget, DifferentialPrivacyDistribution,
+    DifferentialPrivacyStrategy, ZCdpBudget,
+};
+
+/// Sample from the Bernoulli(1/2) distribution.
+///
+/// `sample_bernoulli_standard(rng)` returns numbers distributed as $Bernoulli(1/2)$.
+/// using the given random number generator for base randomness.
+fn sample_bernoulli_standard<R: Rng + ?Sized>(rng: &mut R) -> bool {
+    let mut buffer = [0u8; 1];
+    rng.fill_bytes(&mut buffer);
+    buffer[0] & 1 == 1
+}
+
+/// Sample from the Bernoulli(gamma) distribution, where $gamma /leq 1$.
+///
+/// `sample_bernoulli_frac(gamma, rng)` returns numbers distributed as $Bernoulli(gamma)$.
+/// using the given random number generator for base randomness.
+fn sample_bernoulli_frac<R: Rng + ?Sized>(gamma: &BigURational, rng: &mut R) -> bool {
+    let d = gamma.denom();
+    assert!(!d.is_zero());
+    assert!(gamma <= &BigURational::one());
+
+    // sample uniform biguint in [0,d)
+    let s = rng.gen_range(BigUint::zero()..d.clone());
```

Comment:
What algorithm is used to sample `s` here? We may eventually need to standardize this, so it's worth documenting here what we're expecting `gen_range()` to do.

Reply:
it's using the [uniform sampler on `BigUint`]( https://docs.rs/num-bigint/latest/num_bigint/struct.UniformBigUint.html) provided by the `num_bigint` crate. should i put that in the comment?

Reply:
Bump